core/src/main/java/org/everit/email/store/ri/EmailStoreImpl.java

```diff
@@ -0,0 +1,688 @@
+
+import com.mysema.query.Tuple;
+import com.mysema.query.sql.SQLQuery;
+import com.mysema.query.sql.dml.SQLDeleteClause;
+import com.mysema.query.sql.dml.SQLInsertClause;
+import com.mysema.query.types.Projections;
+
+/**
+ * Reference implementation of {@link EmailStore}.
+ */
+public class EmailStoreImpl implements EmailStore {
+
+  /**
+   * Types of recipient.
+   */
+  private enum RecipientType {
+    BCC, CC, FROM, TO
+  }
+
+  private static final int BUFFER_SIZE = 1024;
+
+  private static final int START_INDEX = 0;
+
+  private final Blobstore blobstore;
+
+  private final QuerydslSupport querydslSupport;
+
+  private final TransactionPropagator transactionPropagator;
+
+  /**
+   * Simple constructor.
+   *
+   * @param querydslSupport
+   *          a {@link QuerydslSupport} instance.
+   * @param transactionPropagator
+   *          a {@link TransactionPropagator} instance.
+   * @param blobstore
+   *          a {@link Blobstore} instance.
+   */
+  public EmailStoreImpl(final QuerydslSupport querydslSupport,
+      final TransactionPropagator transactionPropagator, final Blobstore blobstore) {
+    Objects.requireNonNull(querydslSupport, "The querydslSupport cannot be null!");
+    Objects.requireNonNull(transactionPropagator, "The transactionPropagator cannot be null!");
+    Objects.requireNonNull(blobstore, "The blobstore cannot be null!");
+
+    this.querydslSupport = querydslSupport;
+    this.transactionPropagator = transactionPropagator;
+    this.blobstore = blobstore;
+  }
+
+  private Attachment createAttachment(final Long blobId, final String contentType,
+      final String name) {
+    BlobInputStreamSupplier inputStreamSupplier = null;
+    if (blobId != null) {
+      inputStreamSupplier = new BlobInputStreamSupplier(blobstore, blobId);
+    }
+    Attachment attachment = new Attachment()
+        .withContentType(contentType)
+        .withName(name)
+        .withInputStreamSupplier(inputStreamSupplier);
+    return attachment;
+  }
+
+  private Long createBlob(final byte[] contentBytes) {
+    if ((contentBytes == null) || (contentBytes.length < 1)) {
+      return null;
+    }
+    try (BlobAccessor blobAccessor = blobstore.createBlob()) {
+      blobAccessor.write(contentBytes, 0, contentBytes.length);
+      return blobAccessor.getBlobId();
+    }
+  }
+
+  private Long createBlob(final InputStream inputStream) {
+    if (inputStream == null) {
+      return null;
+    }
+    try (BlobAccessor blobAccessor = blobstore.createBlob()) {
+      int nRead;
+      byte[] data = new byte[BUFFER_SIZE];
+      while ((nRead = inputStream.read(data, 0, data.length)) != -1) {
+        blobAccessor.write(data, 0, nRead);
+      }
+      return blobAccessor.getBlobId();
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void delete(final long storedEmailId) {
+    boolean emailExists = isEmailExists(storedEmailId);
```

Comment:
We do not need to query if an email with an id exists. If we run a DELETE SQL function, we get the number of affected records as a result. If that is 0, then the email did not exist. We want to throw an EmailDoesNotExistException in that case instead of just returning.